staging/src/k8s.io/apiserver/pkg/util/flowcontrol/conc_alloc_test.go

```diff
@@ -0,0 +1,148 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package flowcontrol
+
+import (
+	"math"
+	"math/rand"
+	"sort"
+	"testing"
+)
+
+// TestConcAlloc tests computeConcurrencyAllocation with a bunch of randomly generated cases.
+func TestConcAlloc(t *testing.T) {
+	rands := rand.New(rand.NewSource(1234567890))
+	for i := 0; i < 10000; i++ {
+		probLen := ([]int{0, 1, 2, 3, 4, 6, 9})[rands.Intn(7)]
+		classes := make([]allocProblemItem, probLen)
+		var lowSum, highSum float64
+		var requiredSum int
+		var requiredSumF float64
+		style := "empty"
+		if probLen > 0 {
+			switch rands.Intn(20) {
+			case 0:
+				style = "bound from below"
+				requiredSum = rands.Intn(probLen * 3)
+				requiredSumF = float64(requiredSum)
+				partition32(rands, probLen, requiredSumF, func(j int, x float64) {
+					classes[j].lowerBound = x
+					classes[j].target = x + 2*rands.Float64()
+					classes[j].upperBound = x + 3*rands.Float64()
+					lowSum += classes[j].lowerBound
+					highSum += classes[j].upperBound
+				})
+			case 1:
+				style = "bound from above"
+				requiredSum = rands.Intn(probLen*3) + 1
+				requiredSumF = float64(requiredSum)
+				partition32(rands, probLen, requiredSumF, func(j int, x float64) {
+					classes[j].upperBound = x
+					classes[j].lowerBound = x * math.Max(0, 1.25*rands.Float64()-0.25)
+					classes[j].target = classes[j].lowerBound + rands.Float64()
+					lowSum += classes[j].lowerBound
+					highSum += classes[j].upperBound
+				})
+			default:
+				style = "general"
+				for j := 0; j < probLen; j++ {
+					x := math.Max(0, rands.Float64()*5-1)
+					classes[j].lowerBound = x
+					classes[j].target = x + 2*rands.Float64()
+					classes[j].upperBound = x + 3*rands.Float64()
+					lowSum += classes[j].lowerBound
+					highSum += classes[j].upperBound
+				}
+				requiredSumF = float64(math.Round(float64(lowSum + (highSum-lowSum)*rands.Float64())))
+				requiredSum = int(requiredSumF)
+			}
+		}
+		for rands.Float64() < 0.25 {
+			// Add a class with a target of zero
+			classes = append(classes, allocProblemItem{target: 0, upperBound: rands.Float64() + 0.00001})
+			highSum += classes[probLen].upperBound
+			if probLen > 1 {
+				m := rands.Intn(probLen)
+				classes[m], classes[probLen] = classes[probLen], classes[m]
+			}
+			probLen = len(classes)
+		}
+		allocs, fairProp, err := computeConcurrencyAllocation(requiredSum, classes)
+		var actualSumF float64
+		for _, item := range allocs {
+			actualSumF += item
+		}
+		expectErr := lowSum-requiredSumF > 0.00001 || requiredSumF-highSum > 0.00001
+		if err != nil {
+			if expectErr {
+				t.Logf("For requiredSum=%v, %s classes=%#+v expected error and got %#+v", requiredSum, style, classes, err)
+				continue
+			}
+			t.Fatalf("For requiredSum=%v, %s classes=%#+v got unexpected error %#+v", requiredSum, style, classes, err)
+		} else if expectErr {
```

Comment:
[And with the comment above, this no longer needs an else - because previous branch will always exit.

Reply:
done